xclim/ensembles/_robustness.py

```diff
@@ -0,0 +1,232 @@
+            f"sign_frac from change_significance(ref=ref, fut=fut, test={test}, {kwargs_str})",
+            ref=ref,
+            fut=fut,
+        ),
+    )
+    change_frac.attrs.update(
+        description="Fraction of members showing significant change. " + test_str,
+        units="",
+        test=test,
+        xclim_history=update_history(
+            f"change_frac from change_significance(ref=ref, fut=fut, test={test}, {kwargs_str})",
+            ref=ref,
+            fut=fut,
+        ),
+    )
+    return change_frac, sign_frac
+
+
+def knutti_sedlacek(ref: xr.DataArray, fut: xr.DataArray) -> xr.Dataset:
+    """Robustness metric from Knutti and Sedlacek (2013).
+
+    The robustness metric is defined as R = 1 − A1 / A2 , where A1 is defined
+    as the integral of the squared area between two cumulative density functions
+    characterizing the individual model projections and the multi-model mean
+    projection and A2 is the integral of the squared area between two cumulative
+    density functions characterizing the multi-model mean projection and the historical
+    climate. (Description taken from [knutti2013]_)
+
+    A value of R equal to one implies perfect model agreement. Higher model spread or
+    smaller signal decreases the value of R.
+
+    Parameters
+    ----------
+    ref : DataArray
+      Reference period values along 'time' (nt).
+    fut : DataArray
+      Future ensemble values along 'realization' and 'time' (nr, nt).
+
+    Returns
+    -------
+    R, float
+      The robustness metric.
+
+    References
+    ----------
+    .. [knutti2013] Knutti, R. and Sedláček, J. (2013) Robustness and uncertainties in the new CMIP5 climate model projections. Nat. Clim. Change. doi:10.1038/nclimate1716
+    """
+
+    def _knutti_sedlacek(ref, fut):
+        def diff_cdf_sq_area_int(x1, x2):
+            """Exact integral of the squared area between the non-parametric CDFs of 2 vectors."""
+            y1 = (
+                np.arange(x1.size) + 1
+            ) / x1.size  # Non-parametric CDF on points x1 and x2
+            y2 = (
+                np.arange(x2.size) + 1
+            ) / x2.size  # i.e. y1(x) is the proportion of x1 <= x
+
+            x2_in_1 = np.searchsorted(x1, x2, side="right")  # Where to insert x2 in x1
+            x1_in_2 = np.searchsorted(x2, x1, side="right")  # Where to insert x1 in x2
+
+            x = np.insert(
+                x1, x2_in_1, x2
+            )  # Merge to get all "discontinuities" of the CDF difference
+            y1_f = np.insert(
+                y1, x2_in_1, np.r_[0, y1][x2_in_1]
+            )  # y1 with repeated value (to the right) where x2 is inserted
+            y2_f = np.insert(
+                y2, x1_in_2, np.r_[0, y2][x1_in_2]
+            )  # Same for y2. 0s are prepended where needed.
+
+            # Discrete integral of the squared difference (distance) between the two CDFs.
+            return np.sum(np.diff(x) * (y1_f - y2_f)[:-1] ** 2)
```

Comment:
Have you considered `np.trapz` for the integration ?
`scipy.stats.scoreatpercentile` for the cdf ?

Reply:
or the `quantile` method ?

Reply:
I forgot about "trapz"!  I just tested and speed is pretty much the same as without it. 

As for the other two and AFAIU: What I find nice about this solution is that it is _exact_. If we define the experimental CDF as a stepwise function, this code gives the exact definite integral of the (CDF2 - CDF1)**2. No need to use fancy functions to get a smooth or uniformly-spaced array!

Reply:
Good for me.